sourcegraph/sourcegraph
# Add first version of code insights creation UI

This PR is the first out of three PR for creation UI. 
1. **This Pull Request.**
2. [Introduction page for creation UI](https://github.com/sourcegraph/sourcegraph/pull/20482). https://github.com/sourcegraph/sourcegraph/pull/20482
3. [Creation UI for lang stats insights](https://github.com/sourcegraph/sourcegraph/pull/20548)

This PR is about to add the first implementation of code insight [creation UI.](https://www.figma.com/file/QfeID0EEYMEPdEv8pbGULB/Code-Insights-WIP?node-id=262%3A11)

Compare to the existing approach users can go to the creation page and create code insight by filling form inputs and therefore modify the user or organization settings (the existing approach assumes that the user has to edit that settings configuration manually). You can see the creation form for search-code-insight on a screenshot below.

<img width="988" alt="image" src="https://user-images.githubusercontent.com/18492575/116142873-17ae0680-a6e3-11eb-91c7-20f3b7a1c0a1.png">

**How to test** 
1. Open dev tools and set **_enableCodeInsightCreationUI_** field in local storage to `true`.
2. Go to `/insights` and click configure insight.
3. You must see `/insights/create` page with insight creation form.

**Tech and UX details** 
- For form validation and form management these changes use [react-final-form-hooks](https://github.com/final-form/react-final-form-hooks)
- In case if a user tried to submit the form with invalid inputs we focus first on invalid input and by that force to fill all invalid parts of the creation form. 
- All styles for creation form were written in CSS modules approach. 

**Existing UX problems which will be solved in next iterations** 
- Poor repositories input 
- Support case when a user was included in two or more organizations (add select or some part of UI)
